server/routes/finishOAuthLogin.ts

```diff
@@ -0,0 +1,76 @@
+import { Request, Response, NextFunction } from 'express';
+import { SERVER_URL } from '../config';
+import { NotificationCategories } from '../../shared/types';
+import { factory, formatFilename } from '../../shared/logging';
+import { redirectWithLoginSuccess, redirectWithLoginError } from './finishEmailLogin';
+
+const log = factory.getLogger(formatFilename(__filename));
+
+const finishOAuthLogin = async (models, req: Request, res: Response, next: NextFunction) => {
+  const token = req.query.token;
+  if (!token) {
+    return redirectWithLoginError(res, 'Missing token');
+  }
+
+  // Clear any previous login
+  if (req.user) {
+    req.logout();
+  }
+
+  // Validate login token
+  const tokenObj = await models.LoginToken.findOne({ where: { token, email: '' } });
+  if (!tokenObj || !tokenObj.social_account) {
+    return redirectWithLoginError(res, 'Invalid token');
+  }
+  if (+new Date() >= +tokenObj.expires) {
+    return redirectWithLoginError(res, 'Token expired');
+  }
+
+  // If we are already on the correct domain, continue to log in any
+  // social account associated with the LoginToken. Otherwise redirect
+  const hostname = req.headers['x-forwarded-host'] || req.hostname;
+  if (tokenObj.domain !== hostname) {
+    return res.redirect(`https://${tokenObj.domain}/api/finishOAuthLogin?token=${token}`);
```

Comment:
Could this lead to a redirect loop?

Reply:
It shouldn't, because post-redirect, `hostname === tokenObj.domain`

Reply:
What if the new hostname post-redirect does another redirect that sends us back to this route but from a different domain?

Reply:
Well, this is always a problem with OAuth - if the DNS for the end host is misconfigured, the user ends up in a redirect loop. We can't resolve that without querying for DNS on our side and we probably shouldn't be doing that anyway.